In [31]:
import numpy as np
from statsmodels.stats.weightstats import ztest
from scipy.stats import ttest_ind, ttest_rel

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [2]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [37]:
# з умови задачі 'Середній ріст студентів університету вважається 170 см'
null_mean_value = 170
# alternative='larger' з умови задачі ʼперевірити, чи середній ріст випадкової вибірки студентів істотно більшийʼ
z_stat, p_value = ztest(heights, value=null_mean_value, alternative='larger')
print('Z-stat:', z_stat)
print('P-value:', p_value)
print('Mean:', np.mean(heights))
# робимо висновок
if p_value < 0.05:
  print('Відхиляємо нульому гіпотезу. Cередній зріст студентів з випадкової вибірки істотно більший за 170см ')
else:
  print('Не відхиляємо нульову гіпотезу. Недостатньо статистично значущих доказів, що середній зріст студентів з випадкової вибірки істотно більший за 170см')

Z-stat: 0.8482235371587619
P-value: 0.19815674571091746
Mean: 170.6969696969697
Не відхиляємо нульову гіпотезу. Недостатньо статистично значущих доказів, що середній зріст студентів з випадкової вибірки істотно більший за 170см


#Відповідь

Використовуємо односторонній z-test із параметром alternative='larger', де розглянемо:

*Нульова гіпотеза (H0): μ ≤ 170 (середній ріст студентів не більший за 170 см).*

*Альтернативна гіпотеза (H1): μ > 170 (середній ріст студентів більший за 170 см).*


По результатах z-test **не відхиляємо нульову гіпотезу**. Недостатньо статистично значущих доказів, що середній зріст студентів з випадкової вибірки істотно більший за 170 см при рівні значущості 0.05.

Немає статистично значущих доказів, що середнє вибірки (170.7) відрізняється від гіпотетичного середнього (170).

Отже, згідно проведеного тесту, ми **не можемо стверджувати**, що нова вибірка студентів має середній ріст, який би значно перевищував 170 см.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [41]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [42]:
z_stat, p_value = ztest(group_a_scores, group_b_scores, alternative='two-sided')
print('Z-статистика:', z_stat)
print('P-value:', p_value)
print('Mean group A:', np.mean(group_a_scores))
print('Mean group B:', np.mean(group_b_scores))

if p_value < 0.05:
  print('Відхиляємо нульову гіпотезу. Існує статистична різниця між оцінками груп А та В.')
else:
  print('Не відхиляємо нульову гіпотезу, немає достатніх доказів, що існує статистина різниця між оцінками груп А і В.')

Z-статистика: 2.357354864751502
P-value: 0.01840564855252801
Mean group A: 77.08125
Mean group B: 74.73875000000001
Відхиляємо нульову гіпотезу. Існує статистична різниця між оцінками груп А та В.


#Відповідь
Бачимо, що середня оцінка групи А (77) є вищою за середню оцінку групи B (75).

Протестуємо гіпотезу, що група А, яка отримала нове навчання, отримає оцінки, які суттєво відрізнятимуться від оцінок групи B, а оскільки середнє групи А більше (77), то це показуватиме ефективність нового курсу.

Використовуємо односторонній z-test із параметром alternative='two-sided', де розглянемо:

*Нульова гіпотеза (H0): Немає значущої різниці між оцінками групи А та B.*

*Альтернативна гіпотеза (H1): Різниця існує, при чому ця різниця будь-яка.*

Якщо очікується, що перша вибірка має вищі значення, то у ztest використовується alternative='larger', а якщо різниці не очікується, то alternative='two-sided'.

У варіанті із alternative='two-sided' нульова гіпотеза відхилена. То ж новий курс міг вплинути на різницю в оцінках між групою А та B.



In [43]:
z_stat, p_value = ztest(group_a_scores, group_b_scores, alternative='larger')
print('Z-статистика:', z_stat)
print('P-value:', p_value)
print('Mean group A:', np.mean(group_a_scores))
print('Mean group B:', np.mean(group_b_scores))

if p_value < 0.05:
  print('Відхиляємо нульову гіпотезу. Cередній бал групи А істотно вищий, ніж у групи В.')
else:
  print('Не відхиляємо нульову гіпотезу, немає достатніх доказів, що середній бал групи А істотно вищий, ніж у групи В.')

Z-статистика: 2.357354864751502
P-value: 0.009202824276264005
Mean group A: 77.08125
Mean group B: 74.73875000000001
Відхиляємо нульову гіпотезу. Cередній бал групи А істотно вищий, ніж у групи В.


У варіанті із alternative='larger':

*Нульова гіпотеза (H0): Оцінки в групі А статистично менші чи дорівнюють оцінкам групи B.*

*Альтернативна гіпотеза (H1): Оцінки в групі А статистично вищі за оцінки групи B.*

Нульова гіпотеза відхилена. То ж група А, в якій новий курс, показала статистично кращі результати.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [29]:
selected_group_a_scores = group_a_scores[:5]
print(selected_group_a_scores)
t_stat, p_value = ttest_ind(selected_group_a_scores, group_b_scores, equal_var=False)
print('T-stat:', t_stat)
print('P-value:', p_value)
if p_value < 0.05:
    print('Відхиляємо нульову гіпотезу: є статистично значуща різниця між вибраними оцінками з групи А та оцінками з групи В.')
else:
    print('Не відхиляємо нульову гіпотезу: немає статистично значущої різниці між вибраними оцінками з групи А та оцінками з групи В.')


[78.55, 72.25, 79.88, 75.0, 76.54]
T-stat: 1.1226527301193456
P-value: 0.3010736990344312
Не відхиляємо нульову гіпотезу: немає статистично значущої різниці між вибраними оцінками з групи А та оцінками з групи В.


###Відповідь

Застосувати z-test на 5 записах недоцільно через занадто малий розмір вибірки.

Використовуємо t-test з параметром equal_var=False, тому що очікуємо побачити різну дисперсію (на це мав би вплинути новий курс в групі А) у групах.

На основі перших 5 записів з групи А недостатньо статистичних доказів, що новий курс покращив оцінки в групі А. Порада - продовжувати тестування учнів в групі А, щоб отримати більшу вибірку.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [24]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [47]:
z_stat, p_value = ztest(after_training, before_training, alternative='larger')
print('Z-stat:', z_stat)
print('P-value:', p_value)
print('Mean before:', np.mean(before_training))
print('Mean after:', np.mean(after_training))
if p_value < 0.05:
    print("Відхиляємо нульову гіпотезу: існує значуща зміна в показниках працівників після проходження тренінгу.")
else:
    print("Не відхиляємо нульову гіпотезу: зміни в показниках не є статистично значущими.")

Z-stat: 1.3668871347639466
P-value: 0.08583034256235411
Mean before: 50.885999999999996
Mean after: 54.03600000000001
Не відхиляємо нульову гіпотезу: зміни в показниках не є статистично значущими.


###Відповідь

По результату z-test не відхиляємо нульову гіпотезу: зміни в показниках не є статистично значущими. Тобто недостатьно статистичних доказів, що тренінг покращів результати працівників.

Від'ємне значення z-stat говорить про те, що середнє вибірки знаходиться нижче гіпотетичного середнього, взятого для нульової гіпотези.

Використання z-test для повʼязаних вибірок не є корректним. Натомість можна використати t-test для двох вибірок.

Нульова гіпотеза: "Між вибірками не існує статистично значущої різниці".

In [46]:
t_stat, p_value = ttest_rel(after_training, before_training)
print("T-stat:", t_stat)
print("P-value:", p_value)
if p_value < 0.05:
    print("Відхиляємо нульову гіпотезу: є статистично значуща різниця між показниками працівників до і після курсу")
else:
    print("Не відхиляємо нульову гіпотезу: різниця між показниками працівників не є статистично значущою.")

T-stat: 9.13890009623916
P-value: 4.885873625502449e-10
Відхиляємо нульову гіпотезу: є статистично значуща різниця між показниками працівників до і після курсу


Результати t-test дозволяють відхилити нульову гіпотезу, тобто тренінг дійсно може мати вплив на покращення результатів працівників. Крім того, велике значення T-stat (9.14) вказує на значну різницю між середніми значеннями двох пов'язаних груп, показники значно виросли. Якщо глянути середнє вибірки до, то це було 50.89, а після тренінгу стало 54.04.